### PR TITLE
Rewrite uniq to count

### LINKED ISSUE #49528: Rewrite `uniq` to `count` if a subquery returns distinct values.
```
SELECT uniq(x) FROM (SELECT DISTINCT x ...)
```

can be rewritten to

```
SELECT count() FROM (SELECT DISTINCT x ...)
```

as well as

```
SELECT uniq() FROM (SELECT x ... GROUP BY x)
```

It can apply to all variants of the `uniq` functions except `uniqUpTo`.

The optimization could be implemented on a query pipeline basis.

### PR DESCRIPTION
**Pull request by JackyWoo**
This PR try to close #49528

Optimize `uniq` and its variants(except uniqUpTo) into `count` over subquery.

```SELECT uniq(x ...) FROM (SELECT DISTINCT x ...)```
to
```SELECT count() FROM (SELECT DISTINCT x ...)```

```SELECT uniq(x ...) FROM (SELECT x ... GROUP BY x ...)```
to
 ```SELECT count() FROM (SELECT x ... GROUP BY x ...)```

Note that we rewrite all uniq variants except `uniqUpTo`.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
1. Add rewriter for both old and new analyzer.
2. Add settings `optimize_uniq_to_count` which default is 0.

### Test

- [x] Integration tests

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 8004f37e972c9efc7cfc9d66a1157b020e7b6d89 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52004/8004f37e972c9efc7cfc9d66a1157b020e7b6d89/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        };
+
+        /// Replace uniq of initial query to count
+        if (match_subquery_with_distinct() || match_subquery_with_group_by())
+        {
+            AggregateFunctionProperties properties;
+            auto aggregate_function = AggregateFunctionFactory::instance().get("count", {}, {}, properties);
+            function_node->resolveAsAggregateFunction(std::move(aggregate_function));
+            function_node->getArguments().getNodes().clear();
+            query_node->resolveProjectionColumns({{"count()", function_node->getResultType()}});
```
Comment (JackyWoo) [inline src/Analyzer/Passes/UniqToCountPass.cpp:130]: Comment in `QueryAnalysisPass` shows we should not modify projection columns later. But if not, result of `explain query tree` will have inconsistent result. I am not sure how to handle it, please give a hand.

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        auto & uniq_arguments_nodes = function_node->getArguments().getNodes();
+
+        /// Whether query matches 'SELECT uniq(x ...) FROM (SELECT DISTINCT x ...)'
+        auto match_subquery_with_distinct = [&]() -> bool
+        {
+            if (!subquery_node->isDistinct())
+                return false;
+            /// uniq expression list == subquery group by expression list
+            if (!nodeListEquals(uniq_arguments_nodes, subquery_node->getProjection().getNodes()))
```
Comment (JackyWoo) [inline src/Analyzer/Passes/UniqToCountPass.cpp:104]: When an alias is set in the projection column of a subquery and this alias is used in the uniq function of the outer query, the columns of the subquery and the outer query will not be matched, which will lead to rewriting fail. For example the following SQL:
```
explain query tree SELECT uniq(alias_of_a) FROM (SELECT a as alias_of_a, sum(b) FROM test_rewrite_uniq_to_count GROUP BY a) t settings optimize_uniq_to_count = 1, allow_experimental_analyzer = 1;
```

The reason is that the alias in the outer query `uniq(alias_of_a)` is not expanded. I think we should expand all alias in query after `QueryAnalysis`.

**Comment by JackyWoo** [comment]
@alexey-milovidov May I ask someone review the PR? Thank you.

**Review by yariks5s — APPROVED** [review]
Looks good. @novikd, can you double-check?

**Comment by JackyWoo** [comment]
@yariks5s I pushed a commit https://github.com/ClickHouse/ClickHouse/pull/52004/commits/08409059cc198873ffbf11060bfdabaa0c74f07f to fix rewriting failure for new analyzer in case like:

```
SELECT uniq(alias_of_a) FROM (SELECT DISTINCT a as alias_of_a FROM test_rewrite_uniq_to_count) t
```

The failure is caused by alias of subquery projection nodes, because up-level query just use alias instead of column name.

**Comment by JackyWoo** [comment]
@yariks5s Can you help with the test [error](https://s3.amazonaws.com/clickhouse-test-reports/52004/08409059cc198873ffbf11060bfdabaa0c74f07f/integration_tests__tsan__[6_6].html)?

**Comment by yariks5s** [comment]
> @yariks5s Can you help with the test [error](https://s3.amazonaws.com/clickhouse-test-reports/52004/08409059cc198873ffbf11060bfdabaa0c74f07f/integration_tests__tsan__%5B6_6%5D.html)?

Regarding errors in Kafka, it's ok, we can try again

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        };
+
+        /// Replace uniq of initial query to count
+        if (match_subquery_with_distinct() || match_subquery_with_group_by())
+        {
+            AggregateFunctionProperties properties;
+            auto aggregate_function = AggregateFunctionFactory::instance().get("count", {}, {}, properties);
+
+            function_node->resolveAsAggregateFunction(std::move(aggregate_function));
+            function_node->getArguments().getNodes().clear();
```
Comment (novikd) [inline src/Analyzer/Passes/UniqToCountPass.cpp:180]: Better to write these lines in the opposite order.

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    for (const auto & query_column : query_columns)
+    {
+        auto find = std::find_if(
+            subquery_columns.begin(),
+            subquery_columns.end(),
+            [&](const auto & subquery_column) -> bool
+            {
+                if (auto * column_node = query_column->as<ColumnNode>())
+                {
+                    return subquery_column == column_node->getColumn();
```
Comment (novikd) [inline src/Analyzer/Passes/UniqToCountPass.cpp:66]: Same here

**Review by novikd — COMMENTED** [review]
I recommend reverting this PR because it's unsafe and requires further improvements.

Also, if it's a performance improvement, it requires adding a performance test. It's hard to tell looking into the code if it makes the query's performance better.

File: src/Analyzer/Passes/UniqToCountPass.cpp (hunk 1 above)
Comment (JackyWoo) [reply src/Analyzer/Passes/UniqToCountPass.cpp:183]: @novikd you mean just ignore `resolveProjectionColumns`?

File: src/Analyzer/Passes/UniqToCountPass.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
```
[the remark this answers is not in the document]
Comment (JackyWoo) [reply src/Analyzer/Passes/UniqToCountPass.cpp:94]: @novikd `subquery_column` is projection name from subquery and `column_node` is `ColumnNode` in up-level query node. In my understanding they should be equal. If I was wrong, could you pleade explain more?

**Pull request merged**